<a href="https://colab.research.google.com/github/hugoalfedoputra-ub/ml/blob/main/nn_practice/JST_CH3_Hugo_Alfedo_Putra_225150201111013.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [115]:
import numpy as np

INT_MAX = 2147483647

data = [
    [1,1,1,1],
    [-1,1,-1,-1],
    [1,1,1,-1],
    [1,-1,-1,1]
]

target = [1,1,-1,-1]

lr = 1
ones = np.ones((4,1))

dataset = np.hstack((data, ones))

w = [0 for _ in range(len(dataset[0]))]

# Slides FILKOM pakai bipolar step function
def bip_step(net, th=0):
  if net > th:
    return 1
  elif -1*th <= net and net <= th:
    return 0
  else:
    return -1

# Wikipedia pakai Heaviside step-function (unit step function)
def bin_step(net, th=0):
  return 1 if net>=th else 0

def perceptron(data, target, init, lr=1e-1, delta_rule='filkom', activation=bip_step, stop_early=False, th=0, stop=1e-3):
  # delta can be 'filkom' or otherwise (e.g. 'wiki')
  prev = [INT_MAX for _ in range(len(data[0]))]
  epoch = 0
  validity = [False for i in range(len(data))]
  # print(init)
  while epoch < 10:
    if stop_early and (np.max(np.absolute(np.subtract(init, prev))) < stop or all(validity)):
      break
    prev = init
    for i, row in enumerate(data):
      net = np.dot(row, init)
      f_net = activation(net, th)
      # Check if the f_net is the same as target
      validity[i] = True if f_net == target[i] else False
      delta = [lr * target[i] * row if f_net != target[i] else 0] if delta_rule=='filkom' else [lr * (target[i]-net) * row]
      # Adds the new delta values to the current values
      init = np.add(init, delta)
      # Fixes issues with delta being an array inside an array
      init = init[0] if type(delta[0]) == np.ndarray else init
    # Increment epoch
    epoch += 1
    # Reset validity every epoch
    validity = [False for i in range(len(data))]
    # print(init)

  return init, epoch

Keempat hasil berikut menggunakan delta rule berdasarkan slides FILKOM dan juga berdasarkan Wikipedia (https://en.wikipedia.org/wiki/Perceptron).

Digunakan pula bipolar step function dan binary step function atau dikenal sebagai hstep seperti pada sumber ini: https://en.wikipedia.org/wiki/Heaviside_step_function.

In [116]:
model1, epoch = perceptron(dataset, target, w, lr, delta_rule='filkom')
print("Cara: filkom, aktivasi: bipolar step")
print('w1, w2, w3, w4 =\n', model1[:-1], '\ndan b =', model1[-1], '\ndalam', epoch, 'epoch\n')

model2, epoch = perceptron(dataset, target, w, lr, delta_rule='wiki')
print("Cara: wikipedia, aktivasi: bipolar step")
print('w1, w2, w3, w4 =\n', model2[:-1], '\ndan b =', model2[-1], '\ndalam', epoch, 'epoch\n')

model3, epoch = perceptron(dataset, target, w, lr, delta_rule='filkom', activation=bin_step)
print("Cara: filkom, aktivasi: binary step (hstep)")
print('w1, w2, w3, w4 =\n', model3[:-1], '\ndan b =', model3[-1], '\ndalam', epoch, 'epoch\n')

model4, epoch = perceptron(dataset, target, w, lr, delta_rule='wiki', activation=bin_step)
print("Cara: wikipedia, aktivasi: binary step (hstep)")
print('w1, w2, w3, w4 =\n', model4[:-1], '\ndan b =', model4[-1], '\ndalam', epoch, 'epoch\n')

Cara: filkom, aktivasi: bipolar step
w1, w2, w3, w4 =
 [-2.  2.  0.  2.] 
dan b = 0.0 
dalam 10 epoch

Cara: wikipedia, aktivasi: bipolar step
w1, w2, w3, w4 =
 [26831791. -5728393.  3230775. -5952169.] 
dan b = 17872623.0 
dalam 10 epoch

Cara: filkom, aktivasi: binary step (hstep)
w1, w2, w3, w4 =
 [-14.  10.   6.   6.] 
dan b = -10.0 
dalam 10 epoch

Cara: wikipedia, aktivasi: binary step (hstep)
w1, w2, w3, w4 =
 [26831791. -5728393.  3230775. -5952169.] 
dan b = 17872623.0 
dalam 10 epoch



# Validasi

In [117]:
def test(data, model, activation, target=None, th=0):
  result = []
  for i, row in enumerate(data):
    net = np.dot(row, model)
    f_net = activation(net, th)
    result.append(f_net)
  if target != None:
    print('target:', target)
  print('result:', result)

print("Cara: filkom, aktivasi: bipolar step")
test(dataset, model1, bip_step, target=target)

print("\nCara: wikipedia, aktivasi: bipolar step")
test(dataset, model2, bip_step, target=target)

print("\nCara: filkom, aktivasi: binary step (hstep)")
test(dataset, model1, bin_step, target=target)

print("\nCara: wikipedia, aktivasi: binary step (hstep)")
test(dataset, model2, bin_step, target=target)

Cara: filkom, aktivasi: bipolar step
target: [1, 1, -1, -1]
result: [1, 1, -1, -1]

Cara: wikipedia, aktivasi: bipolar step
target: [1, 1, -1, -1]
result: [1, -1, 1, 1]

Cara: filkom, aktivasi: binary step (hstep)
target: [1, 1, -1, -1]
result: [1, 1, 0, 0]

Cara: wikipedia, aktivasi: binary step (hstep)
target: [1, 1, -1, -1]
result: [1, 0, 1, 1]


# Prediksi

Data berasal dari studi kasus no. 4

In [118]:
data = [
    [-1,-1,1,1,1],
    [-1,-1,-1,-1,1]
]

print("Cara: filkom, aktivasi: bipolar step")
test(data, model1, bip_step)

print("\nCara: wikipedia, aktivasi: bipolar step")
test(data, model2, bip_step)

print("\nCara: filkom, aktivasi: binary step (hstep)")
test(data, model1, bin_step)

print("\nCara: wikipedia, aktivasi: binary step (hstep)")
test(data, model2, bin_step)

Cara: filkom, aktivasi: bipolar step
result: [1, -1]

Cara: wikipedia, aktivasi: bipolar step
result: [-1, -1]

Cara: filkom, aktivasi: binary step (hstep)
result: [1, 0]

Cara: wikipedia, aktivasi: binary step (hstep)
result: [0, 0]


# Diskusi

1. Tampilkan weight terakhir dari hasil pelatihan menggunakan dua weight update

|filkom, bipolar|filkom, hstep|wiki, bipolar|wiki, hstep|
|---|---|---|---|
|`[-2.  2.  0.  2.]`|`[26831791. -5728393.  3230775. -5952169.]`|`[-14.  10.   6.   6.]`|`[26831791. -5728393.  3230775. -5952169.]`|

2. Apakah terdapat kekurangan pada rumus weight pertama dan kedua? Jika ada jelaskan

Sebenarnya dari kedua rumusan weight tersebut tidak dapat kekurangan; di mana rumusan dari slides FILKOM menggunakan hasil setelah diteruskan melalui fungsi aktivasi, sedangkan pada Wikipedia menggunakan hasil weighted sum sebelum diteruskan ke fungsi aktivasi.

3. Jika anda diminta untuk melatih arsitektur perceptron menggunakan 1 juta data dan memprediksi 500 ribu data, rumus weight update mana yang akan anda pilih dan kenapa?

Saya akan memilih menggunakan delta rule milik FILKOM karena secara komputasi lebih sederhana.

4. Lakukan prediksi dengan data berikut pada masing masing rumus weight update perceptron dan tampilkan hasilnya

```
Cara: filkom, aktivasi: bipolar step
result: [1, -1]

Cara: wikipedia, aktivasi: bipolar step
result: [-1, -1]

Cara: filkom, aktivasi: binary step (hstep)
result: [1, 0]

Cara: wikipedia, aktivasi: binary step (hstep)
result: [0, 0]
```

Namun demikian, permasalahan utama berada pada pemilihan nilai learning rate. Saat learning rate = 1, delta rule pada slides FILKOM dapat memprediksi hasil-hasil dengan akurasi 100%, sedangkan menggunakan cara pada Wikipedia hanya memiliki akurasi 25%.

Untuk menjawab pertanyaan sebelumnya, dalam kasus seperti ini, saya akan masih tetap menggunakan cara berdasarkan slides dari FILKOM karena terlihat memiliki akurasi 100%.

Keempat hasil berikut pula menggunakan delta rule dan fungsi aktivasi berdasarkan slides FILKOM dan juga berdasarkan Wikipedia (https://en.wikipedia.org/wiki/Perceptron).

Perbedaannya pada learning rate di mana di-set 0.1 atau `1e-1`

In [119]:
lr = 1e-1

model1, epoch = perceptron(dataset, target, w, lr, delta_rule='filkom')
print("Cara: filkom, aktivasi: bipolar step")
print('w1, w2, w3, w4 =\n', model1[:-1], '\ndan b =', model1[-1], '\ndalam', epoch, 'epoch\n')

model2, epoch = perceptron(dataset, target, w, lr, delta_rule='wiki')
print("Cara: wikipedia, aktivasi: bipolar step")
print('w1, w2, w3, w4 =\n', model2[:-1], '\ndan b =', model2[-1], '\ndalam', epoch, 'epoch\n')

model3, epoch = perceptron(dataset, target, w, lr, delta_rule='filkom', activation=bin_step)
print("Cara: filkom, aktivasi: binary step (hstep)")
print('w1, w2, w3, w4 =\n', model3[:-1], '\ndan b =', model3[-1], '\ndalam', epoch, 'epoch\n')

model4, epoch = perceptron(dataset, target, w, lr, delta_rule='wiki', activation=bin_step)
print("Cara: wikipedia, aktivasi: binary step (hstep)")
print('w1, w2, w3, w4 =\n', model4[:-1], '\ndan b =', model4[-1], '\ndalam', epoch, 'epoch\n')

print("Cara: filkom, aktivasi: bipolar step")
test(dataset, model1, bip_step, target=target)

print("\nCara: wikipedia, aktivasi: bipolar step")
test(dataset, model2, bip_step, target=target)

print("\nCara: filkom, aktivasi: binary step (hstep)")
test(dataset, model1, bin_step, target=target)

print("\nCara: wikipedia, aktivasi: binary step (hstep)")
test(dataset, model2, bin_step, target=target)

Cara: filkom, aktivasi: bipolar step
w1, w2, w3, w4 =
 [-0.2  0.2  0.   0.2] 
dan b = 0.0 
dalam 10 epoch

Cara: wikipedia, aktivasi: bipolar step
w1, w2, w3, w4 =
 [-0.84414579  0.80991057 -0.01054852  0.73874642] 
dan b = -0.02368671098664482 
dalam 10 epoch

Cara: filkom, aktivasi: binary step (hstep)
w1, w2, w3, w4 =
 [-1.4  1.   0.6  0.6] 
dan b = -0.9999999999999999 
dalam 10 epoch

Cara: wikipedia, aktivasi: binary step (hstep)
w1, w2, w3, w4 =
 [-0.84414579  0.80991057 -0.01054852  0.73874642] 
dan b = -0.02368671098664482 
dalam 10 epoch

Cara: filkom, aktivasi: bipolar step
target: [1, 1, -1, -1]
result: [1, 1, -1, -1]

Cara: wikipedia, aktivasi: bipolar step
target: [1, 1, -1, -1]
result: [1, 1, -1, -1]

Cara: filkom, aktivasi: binary step (hstep)
target: [1, 1, -1, -1]
result: [1, 1, 0, 0]

Cara: wikipedia, aktivasi: binary step (hstep)
target: [1, 1, -1, -1]
result: [1, 1, 0, 0]


Terlihat dari hasil validasi:

```
Cara: filkom, aktivasi: bipolar step
target: [1, 1, -1, -1]
result: [1, 1, -1, -1]

Cara: wikipedia, aktivasi: bipolar step
target: [1, 1, -1, -1]
result: [1, 1, -1, -1]

Cara: filkom, aktivasi: binary step (hstep)
target: [1, 1, -1, -1]
result: [1, 1, 0, 0]

Cara: wikipedia, aktivasi: binary step (hstep)
target: [1, 1, -1, -1]
result: [1, 1, 0, 0]
```

di mana kedua delta rule menghasilkan akurasi 100%.

In [120]:
data = [
    [-1,-1,1,1,1],
    [-1,-1,-1,-1,1]
]

print("Cara: filkom, aktivasi: bipolar step")
test(data, model1, bip_step)

print("\nCara: wikipedia, aktivasi: bipolar step")
test(data, model2, bip_step)

print("\nCara: filkom, aktivasi: binary step (hstep)")
test(data, model1, bin_step)

print("\nCara: wikipedia, aktivasi: binary step (hstep)")
test(data, model2, bin_step)

Cara: filkom, aktivasi: bipolar step
result: [1, -1]

Cara: wikipedia, aktivasi: bipolar step
result: [1, -1]

Cara: filkom, aktivasi: binary step (hstep)
result: [1, 0]

Cara: wikipedia, aktivasi: binary step (hstep)
result: [1, 0]


Terlihat sekarang hasil prediksi saat nilai learning rate = 0.1 sama antara perceptron yang menggunakan delta rule pada slides FILKOM maupun Wikipedia.

# Kesimpulan

Tidak ada permasalahan pada delta rule antara slides FILKOM dan Wikipedia. Nilai learning rate memainkan peran penting dalam kovergensi perceptron dan akurasi saat prediksi.